Kernel/Arch/x86/common/Shutdown.cpp

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2022, Liav A. <liavalb@hotmail.co.il>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Arch/Processor.h>
+#include <Kernel/Arch/x86/IO.h>
+#include <Kernel/Arch/x86/common/Shutdown.h>
+
+namespace Kernel {
+
+void qemu_shutdown()
+{
+    // Note: This will invoke QEMU Shutdown, but for other platforms (or emulators),
+    // this has no effect on the system, so we still need to halt afterwards.
+    // We also try the Bochs/Old QEMU shutdown method, if the first didn't work.
```

Comment:
These comments about "try X then try Y, then ..." seem out of place when these functions are dropped into their own file. Should a more generic "write to these ports that don't affect anything when we are not running under XYZ hypervisor" comment be here, with the comments about what order to try them living somewhere higher up the call stack?

Reply:
This will be fixed in #15173.